src/PowerShell/PowerShellManagerPool.cs

```diff
@@ -0,0 +1,63 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using Microsoft.Azure.Functions.PowerShellWorker.Utility;
+
+namespace Microsoft.Azure.Functions.PowerShellWorker.PowerShell
+{
+    using System.Management.Automation;
+
+    /// <summary>
+    /// The PowerShellManager pool for the in-proc concurrency support.
+    /// </summary>
+    internal class PowerShellManagerPool
+    {
+        private readonly ILogger _logger;
+        // Today we don't really support the in-proc concurrency. We just hold an instance of PowerShellManager in this field.
+        private PowerShellManager _psManager;
+
+        /// <summary>
+        /// Constructor of the pool.
+        /// </summary>
+        internal PowerShellManagerPool(ILogger logger)
+        {
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Initialize the pool and populate it with PowerShellManager instances.
+        /// When it's time to really implement this pool, we probably should instantiate PowerShellManager instances in a lazy way.
+        /// Maybe start from size 1 and increase the number of workers as needed.
+        /// </summary>
+        internal void Initialize()
+        {
+            _psManager = new PowerShellManager(_logger);
+        }
+
+        /// <summary>
+        /// Checkout an idle PowerShellManager instance.
+        /// When it's time to really implement this pool, this method is supposed to block when there is no idle instance available.

```

Comment:
I also prefer a lazy way to instantiate workers in the pool. There is a tiny problem though -- when opening a `Runspace`, the `SetModulePath` will be called when creating a `ModuleIntrisic` object, which will change the `PSModulePath` environment variable. Of course, the env variable will soon be updated back to the expected value due to the use of `InitialSessionState.EnvironmentVariables` in this PR, but still, there is a tiny windows where `PSModulePath` is not the expected value. This is a potential race condition, although it might not be a real problem in practice.

Reply:
The pool will have a `max` setting for capacity. When reaching that limit, the `CheckoutIdleWorker` won't be allowed to create a new instance, and thus have to block until an idle worker is available.

Reply:
Yes that makes sense :)